### PR TITLE
Populate Pantheon HUD menu with an AJAX request on hover

### LINKED ISSUE #18: API calls slowing down the backend
We experienced some WordPress admin slowness that was extending from the Pantheon HUD plugin it seems.  It seems if the Pantheon API is struggling, it'll cause some slowness for wp-admin.  

More of an FYI and perhaps discussion starter.

We got a lot of these and had to disable.

```
[01-Mar-2016 18:10:41]  [pool www] pid 12738
script_filename = /srv/bindings/4c91c4cfa37044148799df04d2385d56/code//wp-admin/admin.php
[0x00007fa8ca033aa8] curl_exec() /srv/includes/pantheon.php:212
[0x00007fa8ca0338f8] pantheon_curl() /srv/bindings/4c91c4cfa37044148799df04d2385d56/code/wp-content/plugins/pantheon-hud/inc/class-api.php:113
[0x00007fa8ca033810] fetch_site_data() /srv/bindings/4c91c4cfa37044148799df04d2385d56/code/wp-content/plugins/pantheon-hud/inc/class-api.php:15
[0x00007fa8ca0335d8] __construct() /srv/bindings/4c91c4cfa37044148799df04d2385d56/code/wp-content/plugins/pantheon-hud/inc/class-toolbar.php:28
[0x00007ffcc47b59a0] action_admin_bar_menu() unknown:0
[0x00007fa8ca033410] call_user_func_array() /srv/bindings/4c91c4cfa37044148799df04d2385d56/code/wp-includes/plugin.php:601
[0x00007fa8ca0332e8] do_action_ref_array() /srv/bindings/4c91c4cfa37044148799df04d2385d56/code/wp-includes/admin-bar.php:81
[0x00007ffcc47b5ef0] wp_admin_bar_render() unknown:0
[0x00007fa8ca033150] call_user_func_array() /srv/bindings/4c91c4cfa37044148799df04d2385d56/code/wp-includes/plugin.php:525
[0x00007fa8ca032fb0] do_action() /srv/bindings/4c91c4cfa37044148799df04d2385d56/code/wp-admin/admin-header.php:214
[0x00007fa8ca032db8] +++ dump failed
```

### PR DESCRIPTION
**Pull request by danielbachhuber**
Fixes #18 

pantheonhud

If you'd like to test it locally, drop this code snippet into a mu-plugin to mock a bunch of data:

```php
$_ENV['PANTHEON_ENVIRONMENT'] = 'dev';
$_ENV['PANTHEON_SITE']        = '73cae74a-b66e-440a-ad3b-4f0679eb5e97';
$_ENV['PANTHEON_SITE_NAME']   = 'daniel-pantheon';
$_ENV['php_version']          = '7.3';
define( 'PANTHEON_HUD_PHPUNIT_RUNNING', true );

add_filter(
	'pre_http_request',
	function( $response, $request, $url ) {
		$data_files = [
			'/sites/self/environments/dev/domains'  => 'domains.json',
			'/sites/self/environments/dev/settings' => 'environment-settings.json',
		];
		$path       = wp_parse_url( $url, PHP_URL_PATH );
		if ( ! isset( $data_files[ $path ] ) ) {
			return $response;
		}
		return array(
			'headers'  => array(
				'date'            => 'Thu, 14 Jan 2016 13:46:02 GMT',
				'content-type'    => 'application/json',
				'x-pantheon-host' => 'yggdrasil4ead8a2d.chios.panth.io',
				'server'          => 'TwistedWeb/12.2.0',
			),
			'body'     => file_get_contents( WP_PLUGIN_DIR . '/pantheon-hud/tests/phpunit/data/' . $data_files[ $path ] ), // phpcs:ignore WordPress.WP.AlternativeFunctions
			'response' => array(
				'code'    => 200,
				'message' => 'OK',
			),
			'cookies'  => array(),
			'filename' => null,
		);
	},
	10,
	3
);
```

### CODE REVIEW

**Comment by sugaroverflow** [comment]
Hi, is there an existing multidev site that I can test and verify this works? Or do I need to create my own wp site and install the plugin :D (first time testing something like this) thank you!

File: inc/class-toolbar.php [hunk 1]
```diff
@@ -106,48 +124,49 @@ public function action_admin_bar_menu( $wp_admin_bar ) {
...
+		);
+		$script      = <<<EOT
+jQuery('#wp-admin-bar-pantheon-hud').on('hover', function() {
+	if (jQuery('#wp-admin-bar-pantheon-hud-wp-admin-loading').length) {
+		jQuery.get('{$request_url}', function(data) {
+			jQuery('#wp-admin-bar-pantheon-hud .ab-sub-wrapper').html(data);
+		});
+	}
+});
+EOT;
```
Comment (westonruter) [inline inc/class-toolbar.php]: This presumes that jQuery is on the page, but this is not necessarily the case. In fact, Twenty Nineteen and Twenty Twenty don't use jQuery at all and so it is not enqueued on the page. Core recently improved its handling of the admin bar when jQuery is not present: https://core.trac.wordpress.org/ticket/47069

So this needs to be rewritten to use vanilla JS.

File: inc/class-toolbar.php [hunk 2]
```diff
@@ -106,48 +124,49 @@ public function action_admin_bar_menu( $wp_admin_bar ) {
...
+		wp_add_inline_script( 'admin-bar', $script );
```
Comment (westonruter) [inline inc/class-toolbar.php]: I also have a request to ensure that the script is not removed on AMP pages by flagging it as part of dev mode:

```suggestion
		wp_add_inline_script( 'admin-bar', $script );
add_filter(
			'amp_dev_mode_element_xpaths',
			static function( $xpaths ) {
				$xpaths[] = '//script[ contains( text(), "wp-admin-bar-pantheon-hud" ) ]';
				return $xpaths;
			}
		);
```

Note this code will no longer be needed after https://github.com/ampproject/amp-wp/pull/4001.

File: inc/class-toolbar.php [hunk 3]
```diff
@@ -106,48 +124,49 @@ public function action_admin_bar_menu( $wp_admin_bar ) {
...
+		}
+		$request_url = add_query_arg(
+			array(
+				'action'   => 'pantheon_hud_markup',
+				'_wpnonce' => wp_create_nonce( 'pantheon_hud' ),
+			),
+			admin_url( 'admin-ajax.php' )
+		);
+		$script      = <<<EOT
+jQuery('#wp-admin-bar-pantheon-hud').on('hover', function() {
```
Comment (westonruter) [inline inc/class-toolbar.php]: Also, this needs to take into account keyboard navigation (e.g. it should trigger on `focus` as well).

File: inc/class-toolbar.php (hunk 1 above)
Comment (danielbachhuber) [reply inc/class-toolbar.php]: @westonruter Great suggestion! updated with 5305ce1

File: inc/class-toolbar.php (hunk 3 above)
Comment (danielbachhuber) [reply inc/class-toolbar.php]: Included with 5305ce1

**Comment by danielbachhuber** [comment]
> Hi @danielbachhuber, is there an existing multidev site that I can test and verify this works? Or do I need to create my own wp site and install the plugin :D (first time testing something like this) thank you!

@sugaroverflow Sorry for the late reply here. I wanted to address @westonruter's comments before you checked the changes out.

The plugin isn't auto-deployed to a multidev environment right now. You should be able to install it on any old Pantheon WordPress site though.

**Comment by danielbachhuber** [comment]
Thanks @sugaroverflow !

**Comment by szepeviktor** [comment]
Highly effective!

**Pull request merged**